edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/planners/KubernetesPlanner.cs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Planners
+{
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using k8s;
+    using Microsoft.Azure.Devices.Edge.Agent.Core;
+    using Microsoft.Azure.Devices.Edge.Agent.Docker;
+    using Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Commands;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class KubernetesPlanner<T> : IPlanner
+    {
+        readonly IKubernetes client;
+        readonly ICommandFactory commandFactory;
+        readonly ICombinedConfigProvider<T> combinedConfigProvider;
+        readonly string deviceNamespace;
+        readonly string iotHubHostname;
+        readonly string deviceId;
+
+        public KubernetesPlanner(
+            string deviceNamespace,
+            string iotHubHostname,
+            string deviceId,
+            IKubernetes client,
+            ICommandFactory commandFactory,
+            ICombinedConfigProvider<T> combinedConfigProvider)
+        {
+            this.deviceNamespace = Preconditions.CheckNonWhiteSpace(deviceNamespace, nameof(deviceNamespace));
+            this.iotHubHostname = Preconditions.CheckNonWhiteSpace(iotHubHostname, nameof(iotHubHostname));
+            this.deviceId = Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
+            this.client = Preconditions.CheckNotNull(client, nameof(client));
+            this.commandFactory = Preconditions.CheckNotNull(commandFactory, nameof(commandFactory));
+            this.combinedConfigProvider = Preconditions.CheckNotNull(combinedConfigProvider, nameof(combinedConfigProvider));
+        }
+
+        public async Task<Plan> PlanAsync(
+            ModuleSet desired,
+            ModuleSet current,
+            IRuntimeInfo runtimeInfo,
+            IImmutableDictionary<string, IModuleIdentity> moduleIdentities)
+        {
+            Events.LogDesired(desired);
+            Events.LogCurrent(current);
+            Events.LogIdentities(moduleIdentities);
+
+            // Check that module names sanitize and remain unique.
+            var convertedNames = desired.Modules.Keys.Select(KubeUtils.SanitizeK8sValue).Distinct();
+            if (desired.Modules.Count != convertedNames.Count())
+            {
+                string nameList = desired.Modules.Keys.Join(",");
+                throw new InvalidIdentityException($"Deployment will cause a name collision in Kubernetes namespace, modules: [{nameList}]");
+            }
+
+            // TODO: improve this so it is generic for all potential module types.
+            if (!desired.Modules.Values.All(p => p is IModule<DockerConfig>))
+            {
+                throw new InvalidModuleException("Kubernetes deployment currently only handles type=DockerConfig");
```

Comment:
If you want to make this more generic, you can use typeof(T).FullName instead of hard coding DockerConfig